File: abc357/e/main.py
from collections import deque
N = int(input())
# N, K = map(int, input().split())
A = list(map(int, input().split()))
A = [a-1 for a in A]

F = [[] for i in range(N)]
for i in range(N):
    F[A[i]].append(i)

# print(F)

seen = [0]*N
ans = 0
toadd = [0]*N

for i in range(N):
    if seen[i]:
        continue
    stack = [i]
    now = i
    seen[i] = 1

    while True:
        now = A[now]
        if seen[now]:
            break
        seen[now] = 1
        stack.append(now)

    cycle = [now]
    while True:
        v = stack.pop()
        if v == now:
            break
        cycle.append(v)

    # print(cycle)

    ans += len(cycle)*len(cycle)

    set_cycle = set(cycle)
    for v in cycle:
        Q = deque()
        Q.append(v)
        while Q:
            q = Q.popleft()
            for lq in F[q]:
                if lq in set_cycle:
                    continue
                seen[lq] = 1
                toadd[lq] = toadd[q] +1
                ans += len(cycle) + toadd[lq]
                Q.append(lq)

print(ans)

File: abc357/e/main.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <set>

using namespace std;

int main() {
    int N;
    cin >> N;
    
    vector<int> A(N);
    for (int i = 0; i < N; i++) {
        cin >> A[i];
        A[i]--;  // 0インデックスに調整
    }

    // 隣接リストを作成
    vector<vector<int>> F(N);
    for (int i = 0; i < N; i++) {
        F[A[i]].push_back(i);
    }

    vector<int> seen(N, 0);
    long long ans = 0;

    for (int i = 0; i < N; i++) {
        if (seen[i]) continue;

        vector<int> stack;
        int now = i;
        seen[i] = 1;
        stack.push_back(i);

        // サイクルを検出
        while (true) {
            now = A[now];
            if (seen[now]) break;
            seen[now] = 1;
            stack.push_back(now);
        }

        // サイクルを保存
        vector<int> cycle = {now};
        while (true) {
            int v = stack.back();
            stack.pop_back();
            if (v == now) break;
            cycle.push_back(v);
        }

        ans += (long long)cycle.size() * cycle.size();

        vector<int> toadd(N, 0);

        set<int> set_cycle(cycle.begin(), cycle.end());
        for (int v : cycle) {
            deque<int> Q;
            Q.push_back(v);
            while (!Q.empty()) {
                int q = Q.front();
                Q.pop_front();
                for (int lq : F[q]) {
                    if (set_cycle.count(lq)) continue;
                    seen[lq] = 1;
                    toadd[lq] = toadd[q] + 1;
                    ans += (long long)cycle.size() + toadd[lq];
                    Q.push_back(lq);
                }
            }
        }
    }

    cout << ans << endl;
    return 0;
}
